Support offline-operations of data surface resources

dcdc831 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/kubeedge/kubeedge/blob/master/CONTRIBUTING.md
2. Ensure you have added or ran the appropriate tests for your PR

-->

**What type of PR is this?**

/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind test
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind failing-test
-->


**What this PR does / why we need it**:

In edge computing scenarios, edge nodes are usually under different networks from the cloud and are often offline due to network fluctuations and other issues.  If the edge node is offline, users can only wait for the edge node to come online to perform operations from the cloud if they find any abnormality in the pod or device.

However, the data related to pods and devices are stored locally on the edge nodes, and due to the existence of edged, it is feasible in principle to obtain the data and operate the resources on the edge nodes.  The support of this feature will reduce the network resource consumption of the edge node for debugging and resource adjustment, and avoid the situation that some functions of the KubeEdge platform are not available when the edge node is offline.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #5587 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

wbc6080 reviewed (commented):
Need to make verify CI check pass. You can refer to https://kubeedge.io/docs/developer/make_verify

luomengY commented:
These interfaces need to add tests

dcdc831 commented:
> Need to make verify CI check pass. You can refer to https://kubeedge.io/docs/developer/make_verify

Thank you for your suggestion. I will sort the import and run `make verify` before commit.

wbc6080 reviewed (commented):
Now that #5894 has been merged, it should be able to get the correct DeviceList via MetaServer's API. You can rebase the latest code for testing.
@dcdc831

dcdc831 commented:
I pushed the latest commit with the following major changes:

1. Added tests for the `MetaServer` interfaces of `exec` and `logs`;
2. Added a test for `MetaManager` send messages to DeviceTwin.
3. Sort non-standard imports.
4. Change `fmt` to `klog` in the CLI tool.
5. Rebase the latest code.
6. run `make lint` and `make verify` before push.

luomengY commented on edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go:
```diff
@@ -88,3 +97,139 @@ func (f *Factory) ConfirmUpgrade() http.Handler {
+		stderrStr := queryparams.Get("stderr")
+		ttyStr := queryparams.Get("tty")
+
+		stdin := (stdinStr == trueStr)
```
```suggestion
		stdin := stdinStr == trueStr
```

luomengY commented on edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go:
```diff
@@ -88,3 +97,139 @@ func (f *Factory) ConfirmUpgrade() http.Handler {
+		ttyStr := queryparams.Get("tty")
+
+		stdin := (stdinStr == trueStr)
+		stdout := (stdoutStr == trueStr)
```
```suggestion
		stdout := stdoutStr == trueStr
```

luomengY commented on edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go:
```diff
@@ -88,3 +97,139 @@ func (f *Factory) ConfirmUpgrade() http.Handler {
+
+		stdin := (stdinStr == trueStr)
+		stdout := (stdoutStr == trueStr)
+		stderr := (stderrStr == trueStr)
```
```suggestion
		stderr := stderrStr == trueStr
```

luomengY commented on edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go:
```diff
@@ -88,3 +97,139 @@ func (f *Factory) ConfirmUpgrade() http.Handler {
+		stdin := (stdinStr == trueStr)
+		stdout := (stdoutStr == trueStr)
+		stderr := (stderrStr == trueStr)
+		tty := (ttyStr == trueStr)
```
```suggestion
		tty := ttyStr == trueStr
```

luomengY commented on edge/pkg/metamanager/metaserver/kubernetes/storage/restful/request.go:
```diff
@@ -0,0 +1,100 @@
+	CoreAPIPrefix       = "api"
+	CoreAPIGroupVersion = schema.GroupVersion{Group: "", Version: "v1"}
+	Prefix              = "apis"
+)
```
Delete these unused constants.

Shelley-BaoYue commented on edge/pkg/metamanager/metaserver/common/types.go:
```diff
@@ -4,3 +4,27 @@ type RestartInfo struct {
+	Timestamps                   string `query:"timestamps"`
+}
+
+type ExecInfo struct {
```
Why param. in ExecInfo don't need the format `query:"xxx"` like LogsInfo?

dcdc831 replied:
The `logs` is run on the edged restful API, these tags is used to generate query parameters. `exec` calls the `remoteRuntimeService` methods (`ExecSync` and `Exec`) without generating query parameters.

Shelley-BaoYue commented on edge/pkg/metamanager/metaserver/handlerfactory/ext_handler_test.go:
```diff
@@ -0,0 +1,286 @@
+package handlerfactory
```
Please add copyright for new file

Shelley-BaoYue commented on edge/pkg/metamanager/metaserver/kubernetes/storage/restful/request.go:
```diff
@@ -0,0 +1,96 @@
+/*
```
Please djust the copyright format, like adding blank line between L5-6 and L6-7

Shelley-BaoYue commented on edge/pkg/metamanager/process.go:
```diff
@@ -255,6 +267,12 @@ func (m *metaManager) processUpdate(message model.Message) {
 	case cloudmodules.PolicyControllerModuleName:
 		resp := message.NewRespByMessage(&message, OK)
 		sendToCloud(resp)
+	case modules.MetaManagerModuleName:
```
It may cause confused using `metamanager` module here because we already have a module named `meta`. How about name this module  `metaserver` because it seems the request are from metaserver?   but metaserver is not an independent module, I'm not sure if metaserver is suitable. @fisherxu @luomengY @WillardHu @wbc6080

luomengY replied:
> It may cause confused using `metamanager` module here because we already have a module named `meta`. How about name this module `metaserver` because it seems the request are from metaserver? but metaserver is not an independent module, I'm not sure if metaserver is suitable. @fisherxu @luomengY @WillardHu @wbc6080

I think using a `metaserver` is possible, as the source here comes from the `metaserver`.

luomengY replied:
> It may cause confused using `metamanager` module here because we already have a module named `meta`. How about name this module `metaserver` because it seems the request are from metaserver? but metaserver is not an independent module, I'm not sure if metaserver is suitable. @fisherxu @luomengY @WillardHu @wbc6080

@WillardHu @fisherxu

wbc6080 replied:
> I think using a `metaserver` is possible, as the source here comes from the `metaserver`.
+1

dcdc831 replied:
OK, I added and registered `metaserver` module, and set the msgsource to `metaserver`.

Shelley-BaoYue commented on keadm/cmd/keadm/app/cmd/ctl/describe/device.go:
```diff
@@ -0,0 +1,195 @@
+			if device.Spec.NodeName == nodeName {
+				deviceListFilter.Items = append(deviceListFilter.Items, *device)
+			} else {
+				klog.Infof("can't query device: \"%s\" for node: \"%s\"", device.Name, device.Spec.NodeName)
```
```suggestion
				klog.Errof("can't query device: \"%s\" for node: \"%s\"", device.Name, device.Spec.NodeName)
```

Shelley-BaoYue commented on keadm/cmd/keadm/app/cmd/ctl/describe/device.go:
```diff
@@ -0,0 +1,195 @@
+		return fmt.Errorf("unable to find describer for %v", m)
+	}
+
+	first := true
```
What's the actual effect of `first`?

dcdc831 replied:
`describe` supports printing information about multiple resources. If the `first` is false, blank lines is printed to separate resources.

Shelley-BaoYue commented on keadm/cmd/keadm/app/cmd/ctl/describe/pod.go:
```diff
@@ -0,0 +1,193 @@
+			if pod.Spec.NodeName == nodeName {
+				var apiPod api.Pod
+				if err := k8s_v1_api.Convert_v1_Pod_To_core_Pod(pod, &apiPod, nil); err != nil {
+					klog.Infof("failed to convert pod with err:%v\n", err)
```
```suggestion
					klog.Errof("failed to convert pod with err:%v\n", err)
```
There are several other places using Infof but not Errof. What's the consideration?

dcdc831 commented:
I have some questions about the results of integrationtest. The error occurred in `eventbus` test, and I did not change `eventbus` code.
![image](https://github.com/user-attachments/assets/8104f535-6049-4344-81ab-e281e063011b)

But I executed the 'make Integrationtest' in my local development environment and found that the above error was possible. There are also cases where the test was successfully passed:
![image](https://github.com/user-attachments/assets/f4b3201e-6085-44db-b7d0-22f6436722af)

go version and ginkgo version:
- go version: 1.21.13
- Ginkgo CLI Version: 2.9.5 （Then I run `make integrationtest` several times with Ginkgo CLI Version:2.20.2, this error seems not happening again.）

Shelley-BaoYue commented:
@dcdc831  integration test error is not introduced by this PR. You can ignore it.

dcdc831 commented:
I undo the changes to `metaserver` and keep the msgsource that sends information to `devicetwin` as `metamanager`.

Shelley-BaoYue reviewed (commented):
/lgtm
CIFuzz error can be ignored in this PR, the viaduct migration caused this error.

Shelley-BaoYue commented:
/assign @fisherxu @WillardHu @luomengY

fisherxu commented:
/approve
Thanks for @luomengY @dcdc831 your great work!
